File: CPP/Letter/Analyzer.cpp
//
//  Analyzer.cpp
//  Letter
//
//  Created by Harlan Haskins on 12/27/15.
//  Copyright © 2015 Harlan Haskins. All rights reserved.
//

#include "Analyzer.hpp"

std::vector<std::string> ExpAnalyzer::analyze(std::shared_ptr<Exp> exp) {
    std::vector<std::string> reasons;
    auto funCallExp = dynamic_cast<FunCallExp *>(&*exp);
    if (funCallExp) {
        std::string reason;
        if (!ensureCallExists(*funCallExp, reason)) {
            reasons.push_back(reason);
            return reasons;
        }
        if (!ensureArityMatch(*funCallExp, reason)) {
            reasons.push_back(reason);
            return reasons;
        }
    }
    return reasons;
}

bool ExpAnalyzer::ensureCallExists(FunCallExp exp, std::string &reason) {
    auto func = env.lookupFunc(exp.func);
    if (!func) {
        reason = "Could not find function named \"" + exp.func + "\".";
        return false;
    }
    return true;
}

bool ExpAnalyzer::ensureArityMatch(FunCallExp exp, std::string &reason) {
    auto func = env.lookupFunc(exp.func);
    if (func->arity() != INFINITE_ARITY && func->arity() != exp.args.size()) {
        reason = "Function \"" + exp.func + "\" expects " + std::to_string(func->arity()) + " arguments, but " + std::to_string(exp.args.size()) + " provided.";
        return false;
    }
    return true;
}

std::vector<std::string> FuncAnalyzer::analyze(std::shared_ptr<Func> f) {
    ExpAnalyzer analyzer(this->env);
    UserFunc *userFunc = dynamic_cast<UserFunc *>(&*f);
    if (userFunc) {
        return analyzer.analyze(userFunc->body);
    }
    return std::vector<std::string>();
}

File: src/Letter/Analyzer.hs
module Letter.Analyzer where

import Letter.Core

arity :: FunDef -> Maybe Int
arity (UserFun args _) = Just $ length args
arity (BuiltinFun a _) = a

analyzeExp :: Env -> Exp -> [String]
analyzeExp env funCall@(FunCall _ args) = runChecks env funCall funCallChecks ++ (args >>= analyzeExp env)
analyzeExp env (Let n exp)              = analyzeExp env exp
analyzeExp env _                        = []

analyzeFun :: Env -> (String, FunDef) -> [String]
analyzeFun env (n, UserFun args body) = map (\s -> "In function definition for " ++ n ++ ": \n    " ++ s) (analyzeExp env body)

funCallChecks = [ensureArityMatches]

ensureArityMatches :: Env -> Exp -> Maybe String
ensureArityMatches env (FunCall n args) =
    case findFun env n of
        Nothing     -> Just $ "Could not find function named \"" ++ n ++ "\""
        Just funDef -> case arity funDef of
            Nothing -> Nothing
            Just a -> if a == length args
                       then Nothing
                       else Just $ "Function " ++ n ++ " expects " ++ show a ++ " argument(s), but received " ++ show (length args) ++ "."
ensureArityMatches _ _ = Nothing

runChecks :: Env -> Exp -> [Env -> Exp -> Maybe String] -> [String]
runChecks _ _ [] = []
runChecks env exp (f:fs) =
    case f env exp of
        Nothing -> runChecks env exp fs
        Just error -> error:runChecks env exp fs
